File: abc/abc037/abc037d.py
from sys import setrecursionlimit


def f(i):
    if dp[i] != 0:
        return dp[i]
    y = i // W
    x = i % W
    result = 1
    for dy, dx in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
        ny = y + dy
        nx = x + dx
        if ny < 0 or ny >= H or nx < 0 or nx >= W:
            continue
        if a[ny][nx] <= a[y][x]:
            continue
        result += f(ny * W + nx)
        result %= m
    dp[i] = result
    return result


setrecursionlimit(10 ** 6)
m = 1000000007

H, W = map(int, input().split())
a = [list(map(int, input().split())) for _ in range(H)]

dp = [0] * (H * W)

result = 0
for i in range(H * W):
    result += f(i)
    result %= m
print(result)

File: abc/abc037/abc037_d.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

const (
	m = 1000000007
)

//
var (
	H  int
	W  int
	a  [][]int
	dp [][]int
)

func f(y int, x int) int {
	if dp[y][x] != 0 {
		return dp[y][x]
	}

	result := 1
	for _, d := range [][2]int{[2]int{-1, 0}, [2]int{1, 0}, [2]int{0, -1}, [2]int{0, 1}} {
		ny := y + d[0]
		nx := x + d[1]
		if ny < 0 || ny >= H || nx < 0 || nx >= W {
			continue
		}
		if a[ny][nx] <= a[y][x] {
			continue
		}
		result += f(ny, nx)
		result %= m
	}
	dp[y][x] = result
	return result
}

func main() {
	defer flush()

	H = readInt()
	W = readInt()

	a = make([][]int, H)
	dp = make([][]int, H)
	for y := 0; y < H; y++ {
		a[y] = make([]int, W)
		dp[y] = make([]int, W)
		for x := 0; x < W; x++ {
			a[y][x] = readInt()
		}
	}

	result := 0
	for y := 0; y < H; y++ {
		for x := 0; x < W; x++ {
			result += f(y, x)
			result %= m
		}
	}
	println(result)
}

const (
	ioBufferSize = 1 * 1024 * 1024 // 1 MB
)

var stdinScanner = func() *bufio.Scanner {
	result := bufio.NewScanner(os.Stdin)
	result.Buffer(make([]byte, ioBufferSize), ioBufferSize)
	result.Split(bufio.ScanWords)
	return result
}()

func readString() string {
	stdinScanner.Scan()
	return stdinScanner.Text()
}

func readInt() int {
	result, err := strconv.Atoi(readString())
	if err != nil {
		panic(err)
	}
	return result
}

var stdoutWriter = bufio.NewWriter(os.Stdout)

func flush() {
	stdoutWriter.Flush()
}

func println(args ...interface{}) (int, error) {
	return fmt.Fprintln(stdoutWriter, args...)
}
